packages/emnapi/src/simple-async-operation.ts

```diff
@@ -1,119 +1,51 @@
-    const address = struct >> 2
-    HEAP32[address] = PThread.unusedWorkers.length
-    HEAP32[address + 1] = PThread.runningWorkers.length
+  $emnapiAddSendListener__deps: ['$emnapiGetDynamicCalls'],
+  $emnapiAddSendListener: function (worker: any) {
+    if (worker && !worker._emnapiSendListener) {
+      worker._emnapiSendListener = function _emnapiSendListener (e: any) {
+        const data = ENVIRONMENT_IS_NODE ? e : e.data
+        if (data.emnapiAsyncSend) {
+          emnapiGetDynamicCalls.call_vp(data.emnapiAsyncSend.callback, data.emnapiAsyncSend.data)
+        }
+      }
+      if (ENVIRONMENT_IS_NODE) {
+        worker.on('message', worker._emnapiSendListener)
+      } else {
+        worker.addEventListener('message', worker._emnapiSendListener, false)
+      }
+    }
   },
 
-  // _emnapi_delete_async_work_js: function (_work: number) {}
-
-  _emnapi_on_execute_async_work_js: function (work: number) {
+  _emnapi_async_send__deps: ['$emnapiGetDynamicCalls', '$PThread', '$emnapiAddSendListener'],
+  _emnapi_async_send: function (callback: number, data: number): void {
+    $from64('callback')
+    $from64('data')
     if (ENVIRONMENT_IS_PTHREAD) {
-      $from64('work')
-      postMessage({ emnapiAsyncWorkPtr: work })
-    }
-  }
-})
-
-function _emnapi_queue_async_work_js (work: number): void {
-  $from64('work')
-  const tid = $makeGetValue('work', POINTER_SIZE * 4, '*')
-  if (tid === 0) {
-    emnapiAsyncWorkerQueue.push(work)
-    return
-  }
-  const pthreadValue = PThread.pthreads[tid]
-  const worker = (('worker' in pthreadValue) && ('threadInfoStruct' in pthreadValue)) ? pthreadValue.worker : pthreadValue
-  if (!worker._emnapiAsyncWorkListener) {
-    worker._emnapiAsyncWorkListener = function (this: Worker, e: MessageEvent<any>): any {
-      const data = ENVIRONMENT_IS_NODE ? e : e.data
-      const w: number = data.emnapiAsyncWorkPtr
-      if (w) {
-        const env = $makeGetValue('w', 0, '*')
-        $makeSetValue('w', POINTER_SIZE * 4, '0', '*') // tid
-        const complete = $makeGetValue('w', POINTER_SIZE * 2, '*')
-        if (complete) {
-          const envObject = emnapi.envStore.get(env)!
-          const scope = emnapi.openScope(envObject, emnapi.HandleScope)
-          try {
-            envObject.callIntoModule((_envObject) => {
-              const hint = $makeGetValue('w', POINTER_SIZE * 3, '*')
-              emnapiGetDynamicCalls.call_vpip(complete, env, napi_status.napi_ok, hint)
-            })
-          } catch (err) {
-            emnapi.closeScope(envObject, scope)
-            throw err
-          }
-          emnapi.closeScope(envObject, scope)
+      postMessage({
+        emnapiAsyncSend: {
+          callback,
+          data
         }
-      }
-    }
-    if (ENVIRONMENT_IS_NODE) {
-      worker.on('message', worker._emnapiAsyncWorkListener)
+      })
     } else {
-      worker.addEventListener('message', worker._emnapiAsyncWorkListener, false)
+      setTimeout(() => {
```

Comment:
I saw this in earlier version of the code, but forgot to ask - why are you delaying execution via `setTimeout`? `setTimeout` is guaranteed to add at least 4ms delay per invocation in modern browsers, which can be a lot for high-performance code.

If you don't need the delay, maybe you can just execute the function immediately, but if you do need it, then maybe at least use `Promise.resolve().then(() => ...)` instead which will schedule a microtask instead of a full timeout task.

Reply:
_emnapi_async_send emulates uv_async_send. In my understand, `uv_async_send` actually queue asynchronous callback on event loop, not a microtask based on `Promise`, not sure whether I understand correctly.

Reply:
> why are you delaying execution

I find the official docs said: 

http://docs.libuv.org/en/v1.x/guide/threads.html?highlight=uv_async_send#inter-thread-communication

> Remember: `uv_async_send` is also non-blocking and will return immediately.

So in order to make the JavaScript implementation behavior closer to native Node.js, I think the callback function needs to be called with delay.

Reply:
I look into libuv source code, [`uv__async_spin`](https://github.com/libuv/libuv/blob/v1.43.0/src/unix/async.c#L84) is called both before [calling the async callback](https://github.com/libuv/libuv/blob/v1.43.0/src/unix/async.c#L157-L163) and in [`uv__async_close`](https://github.com/libuv/libuv/blob/v1.43.0/src/unix/async.c#L115-L119), is it more appropriate to use `Promise` than `setTimeout`?

Reply:
I _think_ `Promise` aka microtask would be enough but I'm not sure tbh. At least, if you want a full task, you can use a `MessagePort` trick instead - it was commonly used as a polyfill for experimental `setImmediate` API.

Basically you'd just need to do:

```js
function setImmediate(fn) {
  let { port1, port2 } = new MessageChannel();
  port1.onmessage = fn;
  port2.postMessage(null);
}
```

This waits for a full event loop spin, but would be still much faster than `setTimeout`:

![image](https://user-images.githubusercontent.com/557590/202192445-aac75c72-ab72-4bfd-b5bc-4a3b36450a08.png)